tests/unit/rpc_test.cc

```diff
@@ -1900,3 +1912,94 @@ SEASTAR_TEST_CASE(test_timeout_cancel) {
+            reactor::test::set_long_task_queue_state(reactor::test::long_task_queue_state{
+                .abort_on_too_long_task_queue = true,
+                .max_task_backlog = 500,
+            });
+        }).get();
+
+        constexpr int msg_id = 1;
+        env.register_handler(msg_id, [] (shard_id sending_shard, size_t msgs_to_send, rpc::source<sstring> source) {
+            auto sink = source.make_sink<serializer, sstring>();
+
+            // It is safe to drop the future since the caller awaits for the stream to get closed.
+            (void)seastar::async([sending_shard, msgs_to_send, source, sink] () mutable {
+                auto close_sink = deferred_close(sink);
+                log.info("Handler: send {} messages to shard {}: starting", msgs_to_send, sending_shard);
+                sstring data;
+                data.resize(64, 'x');
+                for (size_t i = 0; i < msgs_to_send; ++i) {
+                    sink(data).get();
+                }
+                sink(sstring()).get();
+                sink.flush().get();
+                close_sink.close_now();
+                log.info("Handler: send {} messages to shard {}: done", msgs_to_send, sending_shard);
+            });
+
+            return sink;
+        }).get();
+
+        size_t msgs_per_shard = 1000000;
+#ifdef SEASTAR_DEBUG
+        msgs_per_shard = 50000;
+#endif
+        env.invoke_on_all([&] (rpc_test_env<>::rpc_test_service& s) {
+            return async([&] {
+                test_rpc_proto::client cl(env.proto(), {}, env.make_socket(), ipv4_addr());
+                auto stop_cl = deferred_stop(cl);
+                auto sink = cl.make_stream_sink<serializer, sstring>(env.make_socket()).get();
+                auto close_sink = deferred_close(sink);
+                auto call = env.proto().make_client<rpc::source<sstring> (shard_id, size_t, rpc::sink<sstring>)>(msg_id);
+                auto source = call(cl, this_shard_id(), msgs_per_shard, sink).get();
+
+                size_t count = 0;
+                for (;;) {
+                    try {
+                        if (auto data = source().get()) {
+                            auto val = std::get<0>(*data);
+                            if (val.size()) {
+                                if (count && !(count % 100000)) {
+                                    log.debug("cl_rep_loop: received {} messages...", count);
+                                }
+                                count++;
+                                continue;
+                            } else {
+                                log.debug("cl_rep_loop: got end-of-stream");
+                                continue;
+                            }
+                        } else {
+                            log.debug("cl_rep_loop: got empty data");
+                            continue;
```

Comment:
Doesn't empty data mark the end-of-stream?

Reply:
Yes, but the connection needs to be drained until we get the closed error to be sure that the other side is done.
Otherwise we'd need a sideband mechanism to await for it.